frontend/openchat-agent/src/utils/promise.ts

```diff
@@ -0,0 +1,23 @@
+export type WaitAllResult<T, E> = {
+    success: T[],
+    errors: E[]
+}
+
+export async function waitAll<T, E>(promises: Promise<T>[]): Promise<WaitAllResult<T, E>> {
+    const results = await Promise.allSettled(promises);
+
+    const success: T[] = [];
+    const errors: E[] = [];
+    for (const result of results) {
+        if (result.status === "fulfilled") {
+            success.push(result.value);
+        } else {
+            errors.push(result.reason as E);
```

Comment:
Seems like we are plucking `E` out of thin air here. Do we need to assume that we know the type of `E` rather than just leaving it as `unknown`?